### DISCUSSION TITLE
Arduino NANO Atmega328 + CNC Sheild V4 + A4899 for LaserGRBL

### DISCUSSION THREAD

**Discussion by uros2002** [arkypita/LaserGRBL#1403] [General]
Hello,

I buy Arduino NANO Atmega328 + CNC Sheild V4 + A4899 for laserGRBL mashine

I need help installing GRBL on it and connecting Limit switches and E-Stop switches
This is the Chinese version that runs on the CH 40 driver.
Which version of GRBL should I install?
How do I set up Laser mode?
How do I connect limit switches and E stop switch?

I am a beginner in this and I would beg for help.

Thanks in advance Dejan

**Comment by StuartB4** [comment]
If you install LaserGRBL then you can flash the Nano board with version 1.1h.
The flashing option is on the Tools menu.
First install the CH340 driver, you can do that with LaserGRBL on the Tools menu.

Then start laserGRBL and make sure the Nano is plugged in the USB port.
Don't click connect but select the Com port. I'll post pictures of what you need to do, if you have'nt already done it. 

If you want to flash the same version but with Homing just select that instead of the one in the second picture below.

**Comment by StuartB4** [comment]
![1](https://user-images.githubusercontent.com/16980026/113946966-7594b000-9801-11eb-8dca-e1b774059ecf.jpg)

![2](https://user-images.githubusercontent.com/16980026/113946976-7a596400-9801-11eb-8c7a-85648a2b32bc.jpg)

![3](https://user-images.githubusercontent.com/16980026/113946977-7deceb00-9801-11eb-8267-a1bcc0d8e007.jpg)

![4](https://user-images.githubusercontent.com/16980026/113946980-81807200-9801-11eb-824f-19d4f508dd2a.jpg)

![5](https://user-images.githubusercontent.com/16980026/113946988-85ac8f80-9801-11eb-95e9-fa24dbfc4b86.jpg)

![6](https://user-images.githubusercontent.com/16980026/113946994-88a78000-9801-11eb-9ffb-3b84f31a36ca.jpg)

![7](https://user-images.githubusercontent.com/16980026/113946998-8ba27080-9801-11eb-9129-c3b7e1ac4137.jpg)

**Comment by StuartB4** [comment]
![image](https://user-images.githubusercontent.com/16980026/113948738-81827100-9805-11eb-81a2-ddb34fdcac38.png)

**Reply by Stevyjb** [reply]
Hello, is the yellow PWM wire connected to  the Z- limit pin correct ?

**Reply by AstroPilot20** [reply]
How do I do that with the laser?
https://de.aliexpress.com/item/1005005398771400.html?spm=a2g0o.order_list.order_list_main.28.27815c5fDfPcLF&gatewayAdapt=glo2deu

**Comment by StuartB4** [comment]
![image](https://user-images.githubusercontent.com/16980026/113950374-1470da80-9809-11eb-9203-ec307c75b1a8.png)

**Comment by uros2002** [comment]
#I did everything as you wrote.

Everything went OK.

Now I have another problem.
When I connect the Arduino Nano and connect it to the laser grbl. after about 1min he starts behaving strangely. It's like losing connection.
The driver is OK don't get lost.

I suspected it was something with CNC Sheild or 4899 drivers so I attached only Arduino NANO, same problem

Attached is a video
https://photos.app.goo.gl/sfpYYCrTsr5SjRzh6
https://photos.app.goo.gl/cC1oCzM1VSjAPFF27


Thank you for your help

**Reply by StuartB4** [reply]
I would try with a different usb lead if you have one.

**Reply by uros2002** [reply]
yes the same thing happens

**Comment by StuartB4** [comment]
It could also be the USB going in to Sleep mode. This may help.
https://www.top-password.com/blog/set-windows-to-power-on-off-usb-devices-in-sleep-mode

Scroll down to where it says Method 2: Enable / Disable USB Selective Suspend Setting

**Reply by uros2002** [reply]
I set it up like from the link, I'll try it now.
thanks

**Reply by uros2002** [reply]
I did not solve the problem behaving the same

**Comment by StuartB4** [comment]
Something else to try is plugging the usb lead in to a usb port on the back of the computer, that goes straight on to the motherboard. 
The ones on the front have extra leads that go from the socket to the motherboard connectors, could be a problem with them.

**Reply by uros2002** [reply]
I also tried on a laptop on the same USB on which my second laser works, which has an Arduino UNO and a CNC Shield V3
He behaves the same

**Comment by uros2002** [comment]
I did not solve the problem behaving the same

**Comment by gmmanonymus111** [comment]
Could you try lowering LaserGRBL's status request to `slow`? Maybe the Nano can't keep up with the constant request messages while it is set to `fast` or above.

**Reply by uros2002** [reply]
The same problem happens when I just connect NANO to lasergrbl.

No launching anything on the lasergrbl, just an established connection

**Reply by gmmanonymus111** [reply]
That's why I suggested lowering the status request. You can't see it, but LaserGRBL send a lot of commands in a short period of time (to show you data about your engraver, like position, free buffer, etc; but the response for that command is hidden from the gcode menu)

**Reply by uros2002** [reply]
How do I do that?

**Reply by gmmanonymus111** [reply]
On LaserGRBL, go to `Grbl`>`Settings`>`protocol`>`Threading mode` and set it to `slow`

**Comment by StuartB4** [comment]
![image](https://user-images.githubusercontent.com/16980026/114166850-26946b00-9926-11eb-87e5-41826aacc837.png)

**Comment by hjsspin** [comment]
I just run into the problem using Lasergrbl with cnc grbl shiled v4(same one as the one OP uses).
It turn out the cnc grbl shield v4 has many rproblems :
1. the mcrostep setting: Ms1 ms2 ms3 is wrongly connect to Gnd (it shoud be Vcc).
this can be justfied by wiring the pins to +5V.
2. The DirX DirY DirZ pin are wrongly connected to D2 D3 D4; StepX Step Y Step Z are wrongly connected to D5 D6 D7, respectively.
The DirX DirY DirZ should be connected to D5 D6 D7; StepX Step Y Step Z to D2 D3 D4, respectively.
I dont know how to work with this issue. Maybe someone experienced can help us.

**Reply by svenhb** [reply]
About the microstep: https://www.instructables.com/Fix-Cloned-Arduino-NANO-CNC-Shield/
New grbl firmware I made for myself: https://github.com/svenhb/GRBL-Plotter/blob/master/Firmware/hex_files/grbl_v1.1f_Servo_switch_dir_step.hex

**Reply by svenhb** [reply]
Sorry Servo-PWM may not work to control laser

**Reply by hjsspin** [reply]
I soldered the MS1 Ms2 Ms3 of X, Y and Z socket together to +5V. This enabled A4988 microstyep setting.
Then I followed the youtube instruction (=> https://www.youtube.com/watch?v=GNVYXwwtoJY) to compile and upload the modified GrblUpload.ino.hex to arduino nano. Finally,  the Step and Dir Pin settings on its firmware are swapped.
I think this  GrblUpload.ino.hex can be renamed and loaded into lasergrbl c# project for easy switching between differnt version of shields.
I would like to add one thing that: during my struggle with the nema 17 motors, I realized a fact that the connector on nema17 has at least two kinds. If the six pins are A_BC_D，(_ means unused), it could be AB shorted, CD shorted; or, AC shorted and BD shorted.
The former requires no swaping B and C wires; while the latter do need to swap between B and C for the motor to work normally.
